From: Chad Dupuis <chad.dupuis@qlogic.com>
Date: Mon, 17 Nov 2014 11:06:31 -0800
Subject: bnx2fc: Cleanup io_req after ABTS timeout.
Patch-Mainline: Not yet
References: bnc#909367,FATE#317542

After the removal of explicit logouts on certain request timeouts we still
need to cleanup up SCSI commands for which the ABTS timed out.  The involves
two steps:

- Send a cleanup request to the firmware
- Return a retryable error back to the SCSI midlayer

Failure to do this was causing some processes to hang waiting for the
request to be returned which resulted in hung_task messages.

CQ: 76818

Signed-off-by: Chad Dupuis <chad.dupuis@qlogic.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/scsi/bnx2fc/bnx2fc_io.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/scsi/bnx2fc/bnx2fc_io.c b/drivers/scsi/bnx2fc/bnx2fc_io.c
index fa421a0..29175cc 100644
--- a/drivers/scsi/bnx2fc/bnx2fc_io.c
+++ b/drivers/scsi/bnx2fc/bnx2fc_io.c
@@ -79,6 +79,11 @@ static void bnx2fc_cmd_timeout(struct work_struct *work)
 					       &io_req->req_flags))) {
 
 				kref_put(&io_req->refcount, bnx2fc_cmd_release);
+				/*
+				 * Cleanup and return original command to
+				 * mid-layer.
+				 */
+				bnx2fc_initiate_cleanup(io_req);
 				spin_unlock_bh(&tgt->tgt_lock);
 
 				return;
-- 
1.8.5.2

